### PR TITLE
RPC Meta device support

### PR DESCRIPTION
**Pull request by pbelevich**
We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76882

### CODE REVIEW

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp
Comment (mrshenli) [inline torch/csrc/distributed/rpc/tensorpipe_agent.cpp:1431]: do we need to add checks somewhere that meta -> not-meta or not-meta -> meta is an error?

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp
Comment (mrshenli) [inline torch/csrc/distributed/rpc/tensorpipe_agent.cpp:1434]: does it make sense to insert meta -> meta by default into the device map instead of patching it here?

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/tensorpipe_agent.cpp:1431]: do we have to specify cpu and meta here? Or will it work if we default to {CPU->CPU, Meta->Meta} if device map is not specified?

File: c10/core/Backend.h
Comment (mrshenli) [inline c10/core/Backend.h:111]: cc @ezyang @jbschlosser

**Review by mrshenli — COMMENTED** [review]
Thanks for adding the support. Is the TensorPipe update intentional? If not, share we do submodule update in a dedicated PR? Thanks!

File: c10/core/Backend.h
Comment (ezyang) [inline c10/core/Backend.h:57]: this looks wrong, Meta should be before NumOptions

**Comment by ezyang** [comment]
What's the intended use case for sending meta tensors over RPC? I have a hard time imagining what its useful for.

**Comment by pbelevich** [comment]
> What's the intended use case for sending meta tensors over RPC? I have a hard time imagining what its useful for.

@ezyang 
1. create a large(really large, I mean REALLY LARGE) model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)
5. profit!

cc @jamesr66a @kwen2501

**Comment by ezyang** [comment]
Is submodule update intentional?

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp
Comment (ezyang) [inline torch/csrc/distributed/rpc/tensorpipe_agent.cpp:116]: and is not portable C++, use &&

File: torch/csrc/distributed/rpc/tensorpipe_utils.cpp
Comment (ezyang) [inline torch/csrc/distributed/rpc/tensorpipe_utils.cpp:95]: This field is never written to, it seems useless?

File: torch/csrc/distributed/rpc/tensorpipe_utils.cpp
Comment (ezyang) [inline torch/csrc/distributed/rpc/tensorpipe_utils.cpp:110]: If the buffer length is zero, do you need to assign it at all? (I'm also not terribly familiar with tensorpipe::Message::Tensor, so I'm not sure how you could be assigning a buffer that TensorPipe ostensibly knows nothing about to it)

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp
Comment (pbelevich) [reply torch/csrc/distributed/rpc/tensorpipe_agent.cpp]: addressed

File: torch/csrc/distributed/rpc/tensorpipe_utils.cpp
Comment (pbelevich) [reply torch/csrc/distributed/rpc/tensorpipe_utils.cpp]: agree, it is reverted

File: torch/csrc/distributed/rpc/tensorpipe_utils.cpp
Comment (pbelevich) [reply torch/csrc/distributed/rpc/tensorpipe_utils.cpp]: agree, it is reverted

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp
Comment (pbelevich) [reply torch/csrc/distributed/rpc/tensorpipe_agent.cpp]: reverted

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp
Comment (pbelevich) [reply torch/csrc/distributed/rpc/tensorpipe_agent.cpp]: reverted

File: c10/core/Backend.h
Comment (pbelevich) [reply c10/core/Backend.h]: reverted

File: torch/csrc/distributed/rpc/tensorpipe_utils.cpp
Comment (ezyang) [inline torch/csrc/distributed/rpc/tensorpipe_utils.cpp:314]: What's going on here?

File: torch/csrc/distributed/rpc/tensorpipe_utils.cpp
Comment (ezyang) [inline torch/csrc/distributed/rpc/tensorpipe_utils.cpp:184]: This seems very wasteful. Why not just manually take care of the `i` increments in the loop below so you don't have to allocate a temporary vector (with all refcount bumps). I would care less if this only affected meta tensor code but this pessimizes all cases.

File: torch/csrc/distributed/rpc/tensorpipe_utils.cpp
Comment (ezyang) [inline torch/csrc/distributed/rpc/tensorpipe_utils.cpp:175]: this seems a bit finely balanced but I'm not sure about a good alternative. It seems like it would be better if the pickler natively understood meta tensors, so that you don't have to stub them out here.

**Comment by ezyang** [comment]
This seems basically OK, deferring final review to whoever is more familiar with the distributed codebase here.

File: torch/csrc/distributed/rpc/tensorpipe_utils.cpp
Comment (pbelevich) [reply torch/csrc/distributed/rpc/tensorpipe_utils.cpp]: reverted, thanks!

File: torch/csrc/distributed/rpc/tensorpipe_utils.cpp
Comment (pbelevich) [reply torch/csrc/distributed/rpc/tensorpipe_utils.cpp]: addressed

File: torch/csrc/distributed/rpc/tensorpipe_utils.cpp
Comment (pbelevich) [reply torch/csrc/distributed/rpc/tensorpipe_utils.cpp:179]: I was thinking about this, but it will require pickler/unpickler to track how many meta/non-meta tensors serialized so far and I decided that it is beyond the scope of this PR

**Review by jamesr66a — APPROVED** [review]
This looks fine to me, I'm going to stamp this pending veto by someone more familiar with the PT RPC+tensorpipe integration. Otherwise, probably fine to merge

@cbalioglu We need a way to transmit type/shape information across the wire for the purposes of shape/type specialized tracing on the RPC target. Currently, the way that is available in PT core (i.e. with the fewest dependency additions) is `MetaTensor`. Maybe we can explore porting to FakeTensor after it has been merged into core and proven out.

File: torch/csrc/distributed/rpc/tensorpipe_utils.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/tensorpipe_utils.cpp:179]: +1 This indeed looks a bit hacky. Shall we create an issue to track this (i.e., adding Meta tensor support to jit pickler/unpickler)? I assume we can treat it as if it is a zero-sized tensor?

**Review by mrshenli — APPROVED** [review]
Stamp to unblock

**Pull request closed without merging**
